src/qudi/hardware/adlink/fastcounter_adlink.py

```diff
@@ -0,0 +1,803 @@
+        """
+        err = AdlinkDataTypes.I16(
+            self._dll.WD_AI_AsyncClear(
+                self._card, ctypes.byref(self._start_pos), ctypes.byref(self._count)
+            )
+        )
+        if self.check_if_error(err, "AsyncClear"):
+            return
+        return
+
+    def is_gated(self):
+        """Check the gated counting possibility.
+
+        @return bool: Boolean value indicates if the fast counter is a gated
+                      counter (TRUE) or not (FALSE).
+        """
+        return True
+
+    def get_binwidth(self):
+        """Returns the width of a single timebin in the timetrace in seconds.
+
+        @return float: current length of a single bin in seconds (seconds/bin)
+        """
+        return self._settings.scan_interval.value / self._clock_freq
+
+    def get_data_trace(self):
+        """Polls the current timetrace data from the fast counter.
+
+        Return value is a numpy array (dtype = int64).
+        The binning, specified by calling configure() in forehand, must be
+        taken care of in this hardware class. A possible overflow of the
+        histogram bins must be caught here and taken care of.
+        If the counter is NOT GATED it will return a tuple (1D-numpy-array, info_dict) with
+            returnarray[timebin_index]
+        If the counter is GATED it will return a tuple (2D-numpy-array, info_dict) with
+            returnarray[gate_index, timebin_index]
+
+        info_dict is a dictionary with keys :
+            - 'elapsed_sweeps' : the elapsed number of sweeps
+            - 'elapsed_time' : the elapsed time in seconds
+
+        If the hardware does not support these features, the values should be None
+        """
+
+        info_dict = {
+            "elapsed_sweeps": self._sweeps,
+            "elapsed_time": time.time() - self._start_time,
+        }
+
+        try:
+            data = np.array(self._measurement_buffer, dtype=np.float64)
+            if self._number_of_averages <= 0:
+                transformed_data = self.transform_raw_data(data)
+                return transformed_data, info_dict
+
+            if np.any(self._last_measurement):
+                self._data_buffer[
+                    self._current_buffer_position % self._number_of_averages
+                ] = data - self._last_measurement
+                self._current_buffer_position += 1
+
+            transformed_data = self.transform_raw_data(
+                np.sum(self._data_buffer, axis=0)
+            )
+            self._last_measurement = data
+
+            return transformed_data, info_dict
+
+        except Exception as e:
+            raise ValueError("Did you invoke the counter settings?") from e
+
+    def transform_raw_data(self, data: np.ndarray):
+        """
+        Method that transform the raw data array to the shape expected by qudi
+        """
+        temp = data.reshape(self._number_of_gates, -1)
+        temp += np.abs(np.min(temp))
```

Comment:
I removed it, however this might cause issues as the built in extraction methods will often require values >=0. This last line here just shifts the values by the minimum amount to ensure values >0.